streams/src/main/java/org/apache/kafka/streams/state/VersionedKeyValueStore.java

```diff
@@ -59,12 +59,15 @@
      * Delete the value associated with this key from the store, at the specified timestamp
      * (if there is such a value), and return the deleted value.
      * <p>
-     * This operation is semantically equivalent to {@link #get(Object, long) #get(key, timestamp)}
-     * followed by {@link #put(Object, Object, long) #put(key, null, timestamp)}.
-     * <p>
      * If the timestamp associated with this deletion is older than the store's grace period
      * (i.e., history retention) relative to the current observed stream time, then the deletion
-     * will not be performed.
+     * will not be performed and {@code null} will be returned.
+     * <p>
+     * This operation is semantically equivalent to {@link #get(Object, long) #get(key, timestamp)}
+     * followed by {@link #put(Object, Object, long) #put(key, null, timestamp)}, with
+     * a caveat that if the deletion timestamp is older than the store's grace period
+     * (i.e., history retention) then the return value is always {@code null}, regardless
+     * of what {@link #get(Object, long) #get(key, timestamp)} would return.
      *
```

Comment:
Add a new paragraph (as discussed in person):
```
<p>
Note: If you want to delete the latest version of a key, you should not do {@link #get(key)} and use the returned {@code validFrom} timestamp to call {@link #delete(key, timestamp) delete(key, validFrom)}. The same applies if you want to delete a specific version that was retrieved via {@link #get(key, timestamp)}. Instead, you should pass a business logic inferred timestamp hat specified when the delete actually happens. For example, it could be the timestamp of the currently processed input record or current stream-time.
```
Something like this -- fee free to reword.

Reply:
Added. LMK what you think.